### PR TITLE
[Testing] Add PhanTypeMismatchDimFetch to static analysis

### PR DESCRIPTION
**Pull request by johnsaigle**
This rule makes sure that when you access an array that the key you're using makes sense. E.g. if you are accessing an element from an array using a non-numeric key, but the array is actually declared to be a string, this will flag that case.

### CODE REVIEW

File: php/libraries/Utility.class.inc [hunk 1]
```diff
@@ -874,11 +874,11 @@ class Utility
      * This is used by create timepoint and start timepoint pages,
      * so it probably shouldn't be in the instrument class.
      *
-     * @param string $dateElement The date element in QuickForm array format.
+     * @param array|string $dateElement The date element in QuickForm array format.
```
Comment (PapillonMcGill) [inline php/libraries/Utility.class.inc]:
```suggestion
     * @param array $dateElement The date element in QuickForm array format.
```

File: modules/next_stage/php/next_stage.class.inc [hunk 2]
```diff
@@ -215,7 +215,7 @@ class Next_Stage extends \NDB_Form
     /**
      * Validate function
      *
-     * @param string $values the value
+     * @param array $values the value
```
Comment (PapillonMcGill) [inline modules/next_stage/php/next_stage.class.inc]: Can also return 'true' (see line 281)

File: php/libraries/File_Upload.class.inc [hunk 3]
```diff
@@ -489,7 +489,7 @@ class File_Upload
      * @param string $dir  The target directory for the file.
      * @param int    $name The proposed name for the file.
      *
-     * @return string New filename
+     * @return array New filename
```
Comment (PapillonMcGill) [inline php/libraries/File_Upload.class.inc]:
```suggestion
     * @return string New filename
```

File: .phan/config.php [hunk 4]
```diff
@@ -24,7 +24,7 @@
 		"PhanUndeclaredMethod",
 		"PhanUndeclaredVariable",
 		"PhanUndeclaredVariableDim",
-        "PhanTypeMismatchDimFetch",
+        "PhanUndeclaredTypeReturnType",
```
Comment (driusan) [inline .phan/config.php]: Bad rebase again?

File: php/libraries/Utility.class.inc [hunk 5]
```diff
@@ -874,11 +874,11 @@ class Utility
...
      * so it probably shouldn't be in the instrument class.
      *
-     * @param string $dateElement The date element in QuickForm array format.
+     * @param array $dateElement The date element in QuickForm array format.
      *
-     * @return True if the date is invalid, false if it's valid.
+     * @return bool True if the date is invalid, false if it's valid.
      */
-    static function _checkDate(string $dateElement): bool
+    static function _checkDate($dateElement): bool
```
Comment (driusan) [inline php/libraries/Utility.class.inc]: why doesn't the signature change to an array, only the doc?

**Comment by johnsaigle** [comment]
See #4320 for information on the checkDate function. It should be deleted. This PR should probably be blocked until that one is merged or else this will just create a conflict.

File: php/libraries/Utility.class.inc [hunk 6]
```diff
@@ -870,6 +870,36 @@ class Utility
...
+        ) {
+            // otherwise, if any of the three elements are empty, return
+            // false b/c date entry has been attempted
+            return false;
+        }
+        return checkdate($dateElement['M'], $dateElement['d'], $dateElement['Y']);
+    }
+
+    /**
+>>>>>>> b479d3bbc... Add PhanTypeMismatchDimFetch
```
Comment (driusan) [inline php/libraries/Utility.class.inc]: Left over conflict marker

**Comment by johnsaigle** [comment]
Travis build failed on 7.3 with 
> 1) CandidateListTestIntegrationTest::testPscidLink
NoSuchElementException: Unable to locate element: {"method":"css selector","selector":"body"}
For documentation on this error, please visit: http://seleniumhq.org/exceptions/no_such_element.html
Build info: version: '2.53.1', revision: 'a36b8b1', time: '2016-06-30 17:37:03'
System info: host: 'f19bd295f6ce', ip: '172.18.0.3', os.name: 'Linux', os.arch: 'amd64', os.version: '4.4.0-101-generic', java.version: '1.8.0_91'
Driver info: driver.version: unknown

Tagging #4394 as related. Restarting Travis now.

File: modules/document_repository/php/document_repository.class.inc [hunk 7]
```diff
@@ -233,7 +233,7 @@ class Document_Repository extends \NDB_Menu_Filter
     /**
      * ParseCategory function
      *
-     * @param string $value the value
+     * @param array $value the value
      *
      * @return array
```
Comment (PapillonMcGill) [inline modules/document_repository/php/document_repository.class.inc]: It look like the return value is a string
Plus, the DB query is not sanitized. Should be change to pselect (in another PR).

File: php/libraries/Utility.class.inc [hunk 8]
```diff
@@ -824,6 +824,35 @@ class Utility
         return $test_names;
     }
 
+    /**
```
Comment (driusan) [inline php/libraries/Utility.class.inc]: Why is this PR adding this function?

Other than that it looks good, given that it's just changing comments and still passing Travis..

**Comment by driusan** [comment]
Actually can you put this on minor before they diverge too much?

File: php/libraries/Utility.class.inc (hunk 8 above)
Comment (johnsaigle) [reply php/libraries/Utility.class.inc]: Rebase error I guess. I'll update it to remove this and also change the base branch.

**Pull request merged**